feat(AWS Cognito): Add Support for Custom Sender Triggers

AustinMathuw opened the pull request:
Hi all!

I saw that #10834 stalled, and I see a need for supporting Custom Sender Triggers for Cognito. I used #10834 as a guide to create my own implementation of the custom sender triggers.

This PR adds support for `CustomSMSSender` and `CustomEmailSender` as [documented by AWS](https://docs.aws.amazon.com/cognito/latest/developerguide/user-pool-lambda-custom-sender-triggers.html).

To address [a comment from the original pull request](https://github.com/serverless/serverless/pull/10834#issuecomment-1068027407), only 1 `kmsKeyId` can be configured per Cognito User Pool.

I have also added various unit tests, as well as an integration test. I did not make new files for these, rather updated the original tests for Cognito User Pools.

Thank you!

Closes: #8889

pgrzesik commented on lib/plugins/aws/package/compile/events/cognito-user-pool.js:
```diff
@@ -141,8 +164,15 @@ class AwsCompileCognitoUserPoolEvents {
       if (functionObj.events) {
         functionObj.events.forEach((event) => {
           if (event.cognitoUserPool && event.cognitoUserPool.existing) {
+            if (!validTriggerSources.includes(event.cognitoUserPool.trigger)) {
```
Why do we need that check?

pgrzesik commented on lib/plugins/aws/package/compile/events/cognito-user-pool.js:
```diff
@@ -259,6 +315,35 @@ class AwsCompileCognitoUserPoolEvents {
+      );
+    }
+
+    if (!kmsKeyId) {
```
Shouldn't it also be enforced on the config schema validation level?

AustinMathuw replied:
This is the only check that I am not sure about. I did some research in how validation is handled and I am not sure how to require `kmsKeyId` only with `CustomSMSSender` and `CustomEmailSender`, but leave it optional for all the other triggers. The attribute (like the `LambdaVersion` attribute) is marked as optional in AWS docs, but is required when using  `CustomSMSSender` and `CustomEmailSender`.

See image:
![image](https://user-images.githubusercontent.com/11302370/178059874-b4351f56-2aa4-4545-bbd3-be14529799b0.png)

AustinMathuw replied:
Further, `kmsKeyId` is only used with `CustomSMSSender` and/or `CustomEmailSender` according to [AWS Docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-cognito-userpool-lambdaconfig.html#cfn-cognito-userpool-lambdaconfig-kmskeyid).

So, `kmsKeyId` is either not used (not even optional) or required (only with `CustomSMSSender` and/or `CustomEmailSender`).

pgrzesik commented on test/integration/aws/cognito-user-pool.test.js:
```diff
@@ -64,6 +75,7 @@ describe('AWS - Cognito User Pool Integration Test', function () {
     return BbPromise.all([
       deleteUserPool(poolExistingSimpleSetup),
       deleteUserPool(poolExistingMultiSetup),
+      deleteUserPool(poolExistingCustomEmailSenderSetup),
```
Did you test that these integration tests are now passing as well? Or how did you validate these changes?

AustinMathuw replied:
I did confirm these integration tests are passing using `AWS_ACCESS_KEY_ID=XXX AWS_SECRET_ACCESS_KEY=xxx npx mocha test/integration/aws/cognito-user-pool.test.js`.

pgrzesik replied:
perfect 👍

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js:
```diff
@@ -41,6 +41,26 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
 
   describe('#newCognitoUserPools()', () => {
+    it('should throw when invalid CUP event is given', () => {
+      awsCompileCognitoUserPoolEvents.serverless.service.functions = {
```
Let's not use such tests - we are slowly migrating away from them and all new tests should be based on `runServerless` utility

AustinMathuw replied:
Sounds good! I did it this way for file consistency. If wanted, I can update the entire file to use runServerless.

pgrzesik replied:
I think for not expanding the scope, let's focus on just new tests and if you'd be interested, please do the refactor for all the other tests in a separate PR, it would be much appreciated 🙇

AustinMathuw commented on lib/plugins/aws/package/compile/events/cognito-user-pool.js:
```diff
@@ -40,6 +43,18 @@ class AwsCompileCognitoUserPoolEvents {
         trigger: { enum: validTriggerSources },
         existing: { type: 'boolean' },
         forceDeploy: { type: 'boolean' },
+        kmsKeyId: {
```
I will check this out.

pgrzesik replied:
oh, in that case maybe let's stay in line with AWS suggestions

pgrzesik commented on docs/providers/aws/events/cognito-user-pool.md:
```diff
@@ -74,26 +74,76 @@ functions:
+          # Can either be KMS Key ARN string or reference to KMS Key Resource ARN (see customEmailSenderFunction below)
+          kmsKeyId: 'arn:aws:kms:eu-west-1:111111111111:key/12345678-9abc-def0-1234-56789abcdef1'
+          # Optional, defaults to 'V1_0'
+          lambdaVersion: 'V1_0'
```
I think there's no point in requiring users to provide this if this is the only possible value - if there's a new one, we can introduce it as an optional field with the default of `V1_0` instead. For now I would just always assume it's `V1_0` - just to clarify, it's mandatory to pass this value to config right now?

AustinMathuw replied:
Curious...

From the AWS docs, the `LambdaVersion` is not a required key, but when omitted, the request fails and AWS says it is a required key. Because of this, I am going to keep the logic the same, but disable the ability to pass in a different lambda version.

If we want to further prune this, we can, but I believe keeping how it is would make it easier to integrate other versions later on.

AustinMathuw commented:
@pgrzesik I added a new commit to address much of your feedback. The only outstanding items from your last review are:

- How we should validate `kmsKeyId ` ([see my comments to the unresolved issue above](https://github.com/serverless/serverless/pull/11201#discussion_r915158581)).
- `lambdaVersion` has been removed as an configurable option, but the logic has not been removed to support new Lambda Versions in the future.

I also made a new integration test that tests creating a new User Pool with a CustomEmailSender trigger, which coincidentally adds a test to check the merge resources functionality as well (1 test now covers two features that were not covered by a test).

pgrzesik commented:
Hey @AustinMathuw - thanks a lot - unfortunately I won't be able to provide feedback in the coming days (I'm on a week-long vacation), so it will have to wait, sorry about that.

AustinMathuw commented:
No worries, I'm off for the weekend as well. I'll await your return!

Thanks and enjoy your time away!

AustinMathuw commented:
Hi @pgrzesik!

I hop you had a great trip! Checking in on this as I will not have access to a computer between 7/29 and 8/16.

Thanks!

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js:
```diff
@@ -13,11 +13,172 @@ const { expect } = chai;
+      fixture: 'function',
+      configExt: serverlessConfigurationExtension,
+      command: 'package',
+    }).then(({ cfTemplate }) => {
```
Let's use async/await

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js:
```diff
@@ -802,4 +963,426 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
     });
   });
+
+  describe('Custom Sender Sources', () => {
```
let's put all new tests based on `runServerless` in a separate `describe`

AustinMathuw replied:
Sounds great!

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js:
```diff
@@ -802,4 +963,426 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
+            configExt: {
+              functions: {
+                first: {
+                  handler: 'index.js',
```
I don't think you have to override handler here

AustinMathuw replied:
To eliminate this, I overrode the `basic` function defined in the `function` fixture, instead of creating a new function.

The function is declared in the `function` fixture here: https://github.com/serverless/serverless/blob/main/test/fixtures/programmatic/function/serverless.yml#L10

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js:
```diff
@@ -802,4 +963,426 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
+            configExt: {
+              functions: {
+                first: {
+                  handler: 'index.js',
```
same here

AustinMathuw replied:
See above.

pgrzesik reviewed (changes requested):
thanks @AustinMathuw for your patience and work on this one 👍 I have some minor comments but overall its looking good, please let me know what do you think.

Additionally, it would be good to add new field to the doc here as well: https://github.com/serverless/serverless/blob/main/docs/providers/aws/guide/serverless.yml.md

AustinMathuw commented on lib/plugins/aws/package/compile/events/cognito-user-pool.js:
```diff
@@ -40,6 +43,8 @@ class AwsCompileCognitoUserPoolEvents {
         existing: { type: 'boolean' },
         forceDeploy: { type: 'boolean' },
+        kmsKeyId: { $ref: '#/definitions/awsKmsArn' },
+        // lambdaVersion: { enum: validTriggerSources }, // Add this back in when more lambda versions are added by AWS
```
Sounds good!

AustinMathuw commented:
@pgrzesik Thanks for the quick reply! I have addressed your feedback and resolved all conversations from your latest requests. Let me know what you think!

Thanks!

pgrzesik commented on test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js:
```diff
@@ -803,3 +952,437 @@ describe('AwsCompileCognitoUserPoolEvents', () => {
   });
 });
+
+describe('AwsCompileCognitoUserPoolEvents - runServerless', () => {
```
Sorry I've missed that earlier, let's follow the convention from other test files where we name the top describe with the path to the test file

pgrzesik reviewed (commented):
Thanks @AustinMathuw - it looks great, I have one more cosmetic change and I think we should be good. 

@mnapoli Do you have anything to add here?

AustinMathuw commented:
Thank you @pgrzesik! I will be creating another pull request right after this one to update the rest of the tests for Cognito Events to use `runServerless`. Your insight here has certainly helped get me prepared for it.

mnapoli reviewed (approved):
@pgrzesik nothing to add, feel free to merge when you think this is ready.

pgrzesik reviewed (approved):
Thanks a lot @AustinMathuw, well done 👏

perpil commented:
I think this change may have broken the documentation on serverless.com. These pages are no longer rendering correctly: [aws serverless.yml](https://www.serverless.com/framework/docs/providers/aws/guide/serverless.yml) and [cognito-user-pool](https://www.serverless.com/framework/docs/providers/aws/events/cognito-user-pool).  It seems isolated to pages touched by this change.  I tried to create a PR, but am not sure what exactly is wrong.  I do see that the language was changed from `yaml` to `yml` on some entries and the change in [guide/serverless.yml.md](https://github.com/serverless/serverless/commit/22802efde1d1721404b4e0d704f7806938183522#diff-8521a6d5b115992a4640eaa29cadd387c5837c50214b170ad1aaa552a47b06f4) might need to be the following, per the cognito-user-pools doc, but honestly am not sure what is making it break.

```yaml
functions:
  presignup:
    # ...
    events:
      - cognitoUserPool:
          pool: MyUserPool
          trigger: PreSignUp
          # Optional, if you're referencing an existing User Pool
          existing: true
          # Optional, for forcing deployment of triggers on existing User Pools
          forceDeploy: true
  customemailsender:
      - cognitoUserPool:
          pool: MyUserPool
          trigger: CustomEmailSender
          # Required, if you're using the CustomSMSSender or CustomEmailSender triggers
          # Can either be KMS Key ARN string or reference to KMS Key Resource ARN
          kmsKeyId: 'arn:aws:kms:eu-west-1:111111111111:key/12345678-9abc-def0-1234-56789abcdef1'
          existing: true
          forceDeploy: true
```

AustinMathuw commented:
Hi @perpil!

I am also not sure what caused this. If it was the `.yaml` to `.yml` change, that was unintentional and should be a quick fix. I don't see anything else that would cause such a break.

On a different note, the example in the doc provided should work due to [this test](https://github.com/serverless/serverless/blob/22802efde1d1721404b4e0d704f7806938183522/test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js#L1287). The restriction is that you cannot configure more than 1 **existing** user pool per function as shown by [this test](https://github.com/serverless/serverless/blob/22802efde1d1721404b4e0d704f7806938183522/test/unit/lib/plugins/aws/package/compile/events/cognito-user-pool.test.js#L773) and [documented here](https://github.com/serverless/serverless/blob/22802efde1d1721404b4e0d704f7806938183522/docs/providers/aws/events/cognito-user-pool.md?plain=1#L177). In the example [here](https://github.com/serverless/serverless/blob/22802efde1d1721404b4e0d704f7806938183522/docs/providers/aws/guide/serverless.yml.md?plain=1#L1134), the function is configuring 2 triggers on a single existing user pool. Unless there is something I am missing, the example should be valid.

@pgrzesik thoughts on documentation breaking?

pgrzesik commented:
Hey @perpil - thanks for reporting and for being vigilant. We're working internally on resolving the problem but it's highly unlikely that it was caused by this PR, it just caused re-render of this section of the docs. There's an issue for breaking docs created here: https://github.com/serverless/serverless/issues/11291
